src/main/java/com/android/volley/CacheDispatcher.java

```diff
@@ -29,7 +33,7 @@
  * refresh are enqueued on the specified network queue for processing
  * by a {@link NetworkDispatcher}.
  */
-public class CacheDispatcher extends Thread {
+public class CacheDispatcher extends Thread implements Request.NetworkRequestCompleteListener {
```

Comment:
Can you push this listener responsibility into a delegate? Best practice is generally to limit scope/capability as much as possible... I bring this up because it is confusing to me to see new methods added to this Thread class that do not execute on this Thread.

Reply:
Do you mean as a nested class in CacheDispatcher?

Reply:
@uhager Yes. Move the listener methods, the mWaitingRequests instance and the `maybeAdd` helper method to a nested class. Then you can just synchronize the methods of that nested class, and it will be very clear what state is being maintained.